### ISSUE TITLE
`separate_by` does not work as expected

### ISSUE THREAD

**Issue by shouya** [zesterer/chumsky#27]
Hi,

Thanks for this wonderful library! I've been using chumsky 0.5.0 on a project and it works great! However, I am baffled by a weird behavior of `separate_by`, which I think may be a potential bug.

Here's a simplified piece of code that reproduces the problem I met:

```rust
use chumsky::prelude::*;

fn main() {
  let digit_list =
    one_of::<char, _, Simple<_>>("123".chars()).separated_by(just(','));
  let letter_list = one_of("abc".chars()).separated_by(just(','));
  let parser = digit_list.clone().or(letter_list);

  // works as expected: this line works as expected
  assert_eq!(digit_list.parse("1,2,3"), Ok(vec!['1', '2', '3']));

  // works as expected: trailing tokens are ignored
  assert_eq!(digit_list.parse("1,2,3X"), Ok(vec!['1', '2', '3']));

  // does not work as expected: the trailing "," is not ignored, where I expect it to be ignored as other trailing tokens
  //
  // expected: Ok(['1', '2', '3'])
  // actual: Err([Simple { span: 6..7, reason: Unexpected, expected: {'2', '1', '3'}, found: None, label: None }])
  assert_eq!(digit_list.parse("1,2,3,"), Ok(vec!['1', '2', '3']));

  // does not work as expected. This result is even weirder. In any case I don't expect an empty list to be returned.
  //
  // expected: Ok(['1', '2', '3'])
  // actual: Ok([])
  assert_eq!(parser.parse("1,2,3,"), Ok(vec!['1', '2', '3']))
}
```

Basically, the trailing separator is not correctly treated as normal trailing tokens and would trigger an error when encountered.

P.S. I knew the existence of the `allow_trailing` option. This example is simplified and doesn't fully represent my use-case, where I cannot allow the trailing separators because I need to leave it out for other part of the program to handle it.

**Comment by shouya** [comment]
Oh, just realized that it has been fixed in https://github.com/zesterer/chumsky/pull/25. I should have checked the master branch before posting it here. I'm closing this issue now.

**Comment by shouya** [comment]
Sorry I have to re-open this issue. Although the problem I described in the OP has all worked as expected in @ryo33's fix, I found that `separate_by` still has a problem when combined with alternation.

Let's look at the code:

```
use chumsky::prelude::*;

fn main() {
  let a_list = just::<char, Simple<_>>('a').separated_by(just(','));
  let b_list = just('b').separated_by(just(','));

  let parser = a_list.or(b_list);

  // works as expected
  assert_eq!(parser.parse("a"), Ok(vec!['a']));

  // expected: ['b']
  // actual: []
  assert_eq!(parser.parse("b"), Ok(vec!['b']));
}
```

**Comment by zesterer** [comment]
> Let's look at the code

This is correct behaviour. Parsers are 'lazy': they'll only parse the input you tell them to parse and no more. Some combinators (`.repeated()`, `.separated_by(...)`, `.or_not()`, etc.) are able to pass no input at all and still produce a valid result.

For example, consider a parser that parses an array:

```rs
let item = expr
    .separated_by(just(','))
    .delimited_by('[', ']');
```

When given the input `[]` this parse succeeds! This only happens because the inner `.separated_by(...)` parser consumed no input (because there were no array items).

Obviously, when you're parsing a full program, you generally don't want the parser to consume only the first valid parse pattern and ignore the rest. To solve this, you can 'force' the parser to parse input right the way to the end using `.then_ignore(end())`.

For example, `just('a').repeated()` will succeed on the input `"bbb"` because it parses no characters and leaves the rest. However, `just('a').repeated().then_ignore(end())` will not because it's expecting the end of the input afterwards.

**Comment by shouya** [comment]
Thanks, your explanation makes total sense to me.

It seems like to me what I need is an option that sort of act like `Repeated::at_least`. For now I'm going to use `try_map` at the end to enforce the number of repetition. But ideally I hope `SeparatedBy` can behave similarly to `Repeated` that has these options (`at_least`, `at_most`) to be play with.

I'll see if I can send a PR for that when I finished what I got on my hand :)
